### PR TITLE
Add ObjC Thread Operational Dataset

### PR DESCRIPTION
**Pull request by sagar-apple**
#### Problem
Need a way to generate a data buffer that represents the Thread Operational Dataset that Matter uses from the individual fields in the Operational Dataset. 

#### Change overview
Added a wrapper object for the Operational Dataset which can provide a NSData representation as well.

#### Testing
How was this tested? (at least one bullet point required)
* If unit tests were added, how do they cover this issue?
** Added a test to create this new type of object wrapper

### CODE REVIEW

File: src/darwin/Framework/CHIP.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -626,7 +638,6 @@
 			isa = XCBuildConfiguration;
 			buildSettings = {
 				CODE_SIGN_STYLE = Automatic;
-				DEVELOPMENT_TEAM = "";
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP.xcodeproj/project.pbxproj:629]: Why this change?

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    if (self = [super init]) {
+        _networkName = networkName;
+        _extendedPANID = extendedPANID;
+        _masterKey = masterKey;
+        _PSKc = PSKc;
+        _channel = channel;
+        _panID = panID;
+        _cppThreadOperationalDataset = chip::Thread::OperationalDataset();
+        [self _populateCppOperationalDataset];
+        if (!_cppThreadOperationalDataset.IsValid(_cppThreadOperationalDataset.AsByteSpan())) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:54]: This `IsValid` check doesn't really make sense.  All it validates is that the raw bytes are made up of `ThreadTLV` objects, which will always be true if the span came from an `AsByteSpan` call on a `chip::Thread::OperationalDataset`.  Just take this part out?

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    return nil;
+}
+
+- (void)_populateCppOperationalDataset
+{
+    _cppThreadOperationalDataset.Clear();
+    _cppThreadOperationalDataset.SetNetworkName([self.networkName cStringUsingEncoding:NSUTF8StringEncoding]);
+
+    uint8_t extendedPanId[chip::Thread::kSizeExtendedPanId];
+    [self.extendedPANID getBytes:&extendedPanId length:chip::Thread::kSizeExtendedPanId];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:69]: So if self.extendedPANID does not actually have `chip::Thread::kSizeExtendedPanId` bytes in it, this will leave part of the buffer un-initialized, right?  Should there be a length check either here or in `initWithNetworkName`?  Similar for the other NSData bits.

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    [self.masterKey getBytes:&masterKey length:chip::Thread::kSizeMasterKey];
+    _cppThreadOperationalDataset.SetMasterKey(masterKey);
+
+    uint8_t PSKc[chip::Thread::kSizePSKc];
+    [self.PSKc getBytes:&PSKc length:chip::Thread::kSizePSKc];
+    _cppThreadOperationalDataset.SetPSKc(PSKc);
+
+    _cppThreadOperationalDataset.SetChannel(self.channel);
+
+    uint16_t * valuePtr = (uint16_t *) [self.panID bytes];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:82]: Why isn't the panID just a `uint16_t`, then?

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface CHIPThreadOperationalDataset : NSObject
+
+extern size_t const CHIPSizeThreadNetworkName;
+extern size_t const CHIPSizeThreadExtendedPanId;
+extern size_t const CHIPSizeThreadMasterKey;
+extern size_t const CHIPSizeThreadMeshLocalPrefix;
+extern size_t const CHIPSizeThreadPSKc;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.h:28]: These should be documented.  In particuluar, the fact that they have something to do with `initWithNetwork` and the sizes of its args?

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+#import "CHIPThreadOperationalDataset.h"
+
+#include "CHIPLogging.h"
+#include <support/Span.h>
+#include <support/ThreadOperationalDataset.h>
+
+size_t const CHIPSizeThreadNetworkName = chip::Thread::kSizeNetworkName;
+size_t const CHIPSizeThreadExtendedPanId = chip::Thread::kSizeExtendedPanId;
+size_t const CHIPSizeThreadMasterKey = chip::Thread::kSizeMasterKey;
+size_t const CHIPSizeThreadMeshLocalPrefix = chip::Thread::kSizeMeshLocalPrefix;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:27]: This one seems to be unused.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm (hunk 2 above)
Comment (sagar-apple) [reply src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:54]: ah, unfortunate. Okay I'll remove it.

File: src/darwin/Framework/CHIP.xcodeproj/project.pbxproj (hunk 1 above)
Comment (sagar-apple) [reply src/darwin/Framework/CHIP.xcodeproj/project.pbxproj:629]: It's unnecessarily defined as "nothing" in the pbxproj file, just took it out because it's not needed for tests.

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm (hunk 6 above)
Comment (sagar-apple) [reply src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:27]: Yeah it is unused. I can remove it.

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm (hunk 4 above)
Comment (sagar-apple) [reply src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:82]: This class should ideally consume Apple's  `THCredentials` object. However this is still in beta, so I'm just conforming to its types for the moment. 
https://developer.apple.com/documentation/threadnetwork/thcredentials

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm (hunk 3 above)
Comment (sagar-apple) [reply src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:71]: So I went and added length checks to all the NSData arguments inside this init helper.

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm (hunk 4 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:90]: I see.  In that case, shouldn't we at least have a `_checkDataLength` here for the panID (should have 2 bytes)?

And a comment explaining why the endianness here is right might be nice.

File: src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        return NO;
+    }
+    _cppThreadOperationalDataset.SetPanId(*valuePtr);
+
+    return YES;
+}
+
+- (BOOL)_checkDataLength:(NSData *)data expectedLength:(size_t)expectedLength
+{
+    if (data.length < expectedLength) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/CHIPThreadOperationalDataset.mm:101]: Should this perhaps be an equality check?  Or is it really OK to silently drop part of the incoming data if it's too long?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]
